File: system/apps/chat/frontend/src/presence.ts
/**
 * The chat page's presence reports (see `presence.py`): `hidden` once the shell has handed
 * the page its handshake, `visible` on `shell:shown`, `hidden` on `shell:hidden`, `closed` on
 * `pagehide`, and a heartbeat of the current state every minute so a page that vanished
 * without its `pagehide` stops counting on its own. Only the chat's own page reports (a
 * subagent view's reports would overwrite it: one report per chat and client is kept). The
 * OOM prioritizer reads the aggregate.
 */

import { apiUrl } from "@imbue/workspace-ui/src/base-path";

export type PresenceState = "visible" | "hidden" | "closed";

// Matches PRESENCE_HEARTBEAT_SECONDS in presence.py.
const HEARTBEAT_MS = 60_000;

let heartbeat: ReturnType<typeof setInterval> | null = null;
let currentState: PresenceState = "hidden";
let reportingChatId: string | null = null;
let reportingClientId: string | null = null;

function post(chatId: string, clientId: string, state: PresenceState): void {
  // keepalive lets the closed report leave with the page on pagehide.
  void fetch(apiUrl(`/api/chats/${encodeURIComponent(chatId)}/presence`), {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ client_id: clientId, state }),
    keepalive: true,
  }).catch(() => {
    // Best-effort: the next heartbeat corrects a dropped report.
  });
}

/** Start reporting for this page's chat as `clientId`; a second call re-keys the reports. */
export function startPresenceReporting(chatId: string, clientId: string, initialState: PresenceState): void {
  reportingChatId = chatId;
  reportingClientId = clientId;
  currentState = initialState;
  post(chatId, clientId, initialState);
  if (heartbeat === null) {
    heartbeat = setInterval(() => {
      if (reportingChatId !== null && reportingClientId !== null && currentState !== "closed") {
        post(reportingChatId, reportingClientId, currentState);
      }
    }, HEARTBEAT_MS);
  }
}

/** Report a change of state; a no-op until reporting has started. */
export function reportPresence(state: PresenceState): void {
  currentState = state;
  if (reportingChatId === null || reportingClientId === null) return;
  post(reportingChatId, reportingClientId, state);
}

export function currentPresenceState(): PresenceState {
  return currentState;
}

File: system/apps/chat/imbue/chat/presence.py
"""Per-chat, per-client presence: which clients have a chat's page open, and whether it is showing.

The chat page reports its own presence to the chat app (contracts.md section 10):
``hidden`` once the shell has handed it its handshake,
``visible`` on ``shell:shown``, ``hidden`` again on ``shell:hidden``, ``closed`` on
``pagehide``, and a heartbeat of its current state every minute. Only the chat's own page
reports: a subagent view is a second page of the same chat in the same client, and one
standing report per chat and client is kept here, so its reports would overwrite the chat
page's. The OOM prioritizer reads the aggregate: a chat is *open* while any client has an
unexpired report, and *visible* while any client's last report says so. A report expires
after ten minutes, so a page that vanished without its ``pagehide`` (a crashed tab, a lost
laptop) stops counting on its own.
"""

import threading
import time
from collections.abc import Callable
from enum import auto
from typing import Final

from pydantic import Field
from pydantic import PrivateAttr

from imbue.chat.primitives import ChatId
from imbue.imbue_common.enums import LowerCaseStrEnum
from imbue.imbue_common.frozen_model import FrozenModel
from imbue.imbue_common.mutable_model import MutableModel

# A report that has not been refreshed for this long no longer counts. Ten heartbeats,
# so a page has to miss every one before its chat reads as closed.
PRESENCE_EXPIRY_SECONDS: Final[float] = 600.0

# How often the page re-reports its current state.
PRESENCE_HEARTBEAT_SECONDS: Final[float] = 60.0


class PresenceState(LowerCaseStrEnum):
    """What one client last said about one chat's page (a wire value of the presence route)."""

    VISIBLE = auto()
    HIDDEN = auto()
    CLOSED = auto()


class PresenceReport(FrozenModel):
    """The body of ``POST /api/chats/<id>/presence``."""

    client_id: str = Field(min_length=1, description="The reporting client, as the shell's handshake named it")
    state: PresenceState = Field(description="The page's state in that client")


class _ClientPresence(FrozenModel):
    """One client's standing report about one chat."""

    state: PresenceState = Field(description="visible or hidden; a closed report deletes the record instead")
    reported_at: float = Field(description="Wall-clock epoch seconds of the report, for expiry")


class PresenceTracker(MutableModel):
    """Holds every client's last presence report per chat and answers the aggregate questions.

    Thread-safe: reports arrive on request threads while the prioritizer's sweep reads.
    ``clock`` supplies wall-clock epoch seconds so tests can advance time explicitly.
    """

    model_config = {"arbitrary_types_allowed": True, "extra": "forbid", "frozen": False}

    clock: Callable[[], float] = Field(default=time.time, frozen=True, description="Wall-clock epoch seconds")
    expiry_seconds: float = Field(
        default=PRESENCE_EXPIRY_SECONDS, frozen=True, description="How long an unrefreshed report counts"
    )

    _presence_by_client_by_chat: dict[ChatId, dict[str, _ClientPresence]] = PrivateAttr(default_factory=dict)
    _lock: threading.Lock = PrivateAttr(default_factory=threading.Lock)

    def record(self, chat_id: ChatId, client_id: str, state: PresenceState) -> None:
        """Replace ``client_id``'s standing report about ``chat_id``; ``CLOSED`` drops it."""
        with self._lock:
            by_client = self._presence_by_client_by_chat.setdefault(chat_id, {})
            if state is PresenceState.CLOSED:
                by_client.pop(client_id, None)
                if not by_client:
                    del self._presence_by_client_by_chat[chat_id]
                return
            by_client[client_id] = _ClientPresence(state=state, reported_at=self.clock())

    def forget_chat(self, chat_id: ChatId) -> None:
        with self._lock:
            self._presence_by_client_by_chat.pop(chat_id, None)

    def is_open(self, chat_id: ChatId) -> bool:
        """Whether any client holds an unexpired report about the chat, visible or hidden."""
        return len(self._live_reports(chat_id)) > 0

    def is_visible(self, chat_id: ChatId) -> bool:
        """Whether any client's unexpired last report says the chat is showing."""
        return any(report.state is PresenceState.VISIBLE for report in self._live_reports(chat_id))

    def open_chat_ids(self) -> set[ChatId]:
        with self._lock:
            chat_ids = list(self._presence_by_client_by_chat)
        return {chat_id for chat_id in chat_ids if self.is_open(chat_id)}

    def visible_chat_ids(self) -> set[ChatId]:
        with self._lock:
            chat_ids = list(self._presence_by_client_by_chat)
        return {chat_id for chat_id in chat_ids if self.is_visible(chat_id)}

    def _live_reports(self, chat_id: ChatId) -> list[_ClientPresence]:
        """The unexpired reports about ``chat_id``, dropping the expired ones as they are found."""
        now = self.clock()
        with self._lock:
            by_client = self._presence_by_client_by_chat.get(chat_id)
            if by_client is None:
                return []
            expired_client_ids = [
                client_id for client_id, report in by_client.items() if now - report.reported_at > self.expiry_seconds
            ]
            for client_id in expired_client_ids:
                del by_client[client_id]
            if not by_client:
                del self._presence_by_client_by_chat[chat_id]
            return list(by_client.values())
